ARCs/arc-0056.md

```diff
@@ -0,0 +1,381 @@
+      local: { [name: string]: StorageMap };
+      box: { [name: string]: StorageMap };
+    };
+  };
+  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
+  bareActions: {
+    /** OnCompletes this method allows when appID === 0 */
+    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
+    /** OnCompletes this method allows when appID !== 0 */
+    call: (
+      | "NoOp"
+      | "OptIn"
+      | "CloseOut"
+      | "ClearState"
+      | "UpdateApplication"
+      | "DeleteApplication"
+    )[];
+  };
+  /** Information about the TEAL programs */
+  sourceInfo?: {
+    /** Approval program information */
+    approval: SourceInfo[];
+    /** Clear program information */
+    clear: SourceInfo[];
+  };
+  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23 */
+  source?: {
+    /** The approval program */
+    approval: string;
+    /** The clear program */
+    clear: string;
+  };
+  /** The compiled bytecode for the application. MUST be omitted if included as part of ARC23 */
+  byteCode?: {
+    /** The approval program */
+    approval: string;
+    /** The clear program */
+    clear: string;
+  };
+  /** Information used to get the given byteCode and/or PC values in sourceInfo. MUST be given if byteCode or PC values are present */
+  compilerInfo?: {
+    /** The name of the compiler */
+    compiler: "algod" | "puya";
+    /** Compiler version information */
+    compilerVersion: {
+      major: number;
+      minor: number;
+      patch: number;
+      commit?: string;
+    };
+  };
+  /** ARC-28 events that MAY be emitted by this contract */
+  events?: Array<Event>;
+  /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) to their respecive types and values (if applicable) */
+  templateVariables?: {
+    [name: string]: {
+      /** The type of the template variable */
+      type: ABIType | AVMBytes | StructName;
+      /** If given, the the base64 encoded value used for the given app/program */
+      value?: string;
+    };
+  };
+  /** The scratch variables used during runtime */
+  scratchVariables?: {
+    [name: string]: {
+      slot: number;
+      type: ABIType | AVMBytes | StructName;
+    };
+  };
+}
+```
+
+### Method Interface
+
+Every method in the contract is described via a `Method` interface. This interface is an extension of the one defined in [ARC-4](./arc-0004.md).
+
+```ts
+/** Describes a method in the contract. This interface is an extension of the interface described in ARC-4 */
+interface Method {
+  /** The name of the method */
+  name: string;
+  /** Optional, user-friendly description for the method */
+  desc?: string;
+  /** The arguments of the method, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
```

Comment:
Something I just noticed that confused me was that here we have `struct`, but for the `keyType`/`valueType` in storage we don't have that separation, we just have the struct name. That's quite confusing. Should we include `keyStruct` and `valueStruct` to avoid this confusion?

Reply:
The reason we do this is because the `type` field needs to be backwards compatible with ARC4, but `keyType` and `valueType` don't.

Reply:
Yeah I eventually figured that out. I think we should change the descriptions of all of these fields to be explicit about it all so if you consume arc-56 it’s more obvious. The fact you sometimes use struct and sometimes don’t is a bit confusing.